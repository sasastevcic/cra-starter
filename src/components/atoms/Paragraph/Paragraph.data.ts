import type { ReactNode, ElementType, ComponentPropsWithoutRef } from 'react';
import { ColorType } from '../../../styles/config/theme';
import type { FontWeight, TextAlign, TextTransform } from '../../../styles/config/variables';

export enum ParagraphType {
	Large = 'large',
	Normal = 'normal',
	Small = 'small',
}

export interface ParagraphProps extends ComponentPropsWithoutRef<'p'> {
	as?: ElementType;
	type?: ParagraphType;
	fontWeight?: FontWeight;
	textTransform?: TextTransform;
	textAlign?: TextAlign;
	color?: ColorType;
	children: ReactNode;
}

export interface StyledParagraphProps {
	$type: ParagraphType;
	$fontWeight?: FontWeight;
	$textTransform?: TextTransform;
	$textAlign?: TextAlign;
	$color?: ColorType;
}
