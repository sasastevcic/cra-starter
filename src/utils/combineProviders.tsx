/* eslint-disable @typescript-eslint/no-explicit-any */
import { ComponentProps, ComponentType, FunctionComponent, ReactNode } from 'react';

type Providers = Array<[ComponentType<any>, ComponentProps<any>?]>;

export const combineProviders = (
	providers: Providers,
): FunctionComponent<{ children: ReactNode }> =>
	providers.reduce(
		(AccumulatedProviders, [Provider, props = {}]) =>
			({ children }) =>
				(
					<AccumulatedProviders>
						<Provider {...props}>{children}</Provider>
					</AccumulatedProviders>
				),
		({ children }) => <>{children}</>,
	);
