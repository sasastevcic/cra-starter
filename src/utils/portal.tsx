import { ReactNode, ReactPortal, useMemo } from 'react';
import { createPortal } from 'react-dom';

interface Portal {
	portalId: string;
	children: ReactNode;
}

const Portal = ({ portalId, children }: Portal): ReactPortal => {
	const portalElement = useMemo(() => document.getElementById(portalId), [portalId]);

	if (!portalElement) {
		throw new Error(`Element not found! Element id: ${portalId}`);
	}

	return createPortal(children, portalElement);
};

export default Portal;
