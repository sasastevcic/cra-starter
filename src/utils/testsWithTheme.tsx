/* eslint-disable @typescript-eslint/ban-ts-comment */
import { ReactElement, ComponentType, ReactNode } from 'react';
import { render, RenderOptions, RenderResult } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { theme } from '../styles/config/theme';

export const componentWithTheme = (component: ReactNode): ReactElement => (
	// @ts-ignore
	<ThemeProvider theme={theme}>{component}</ThemeProvider>
);

export const renderWithTheme = (component: ReactElement, options?: RenderOptions): RenderResult => {
	const wrapper: ComponentType<{ children: ReactNode }> = ({ children }) =>
		componentWithTheme(children);

	return render(component, { wrapper, ...options });
};
