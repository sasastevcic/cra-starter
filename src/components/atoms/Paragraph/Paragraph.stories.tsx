import { ReactElement } from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';
import { FONT_SIZE_REGEX } from '../../../constants/regex';
import { typeStyle } from '../../../styles/config/typeStyles';
import { Table as GenericTable } from '../../../styles/storybook/Storybook';
import { StyledRow, StyledData } from '../../../styles/storybook/Storybook.styles';
import ParagraphComponent from './index';
import { ParagraphProps, ParagraphType } from './Paragraph.data';

const Table = (props: ParagraphProps): ReactElement => {
	const tableHeadings = ['Type', 'Size', 'Preview'];

	const rows = Object.entries(ParagraphType).map(([name, type]) => {
		const fontSizesRegex = typeStyle[type].join('').match(FONT_SIZE_REGEX);

		const fontSizes = fontSizesRegex?.filter((fontSize) => !fontSize.includes('calc')) ?? [];

		if (fontSizes.length > 1) {
			tableHeadings[1] = 'Size (Fluid)';
		}

		return (
			<StyledRow key={name}>
				<StyledData>
					<ParagraphComponent>{name}</ParagraphComponent>
				</StyledData>
				<StyledData>
					<ParagraphComponent>{fontSizes.join(' - ')}</ParagraphComponent>
				</StyledData>
				<StyledData>
					<ParagraphComponent type={type} {...props}>
						Paragraph
					</ParagraphComponent>
				</StyledData>
			</StyledRow>
		);
	});

	return <GenericTable headings={tableHeadings}>{rows}</GenericTable>;
};

export default {
	title: 'Types/Paragraphs',
	component: ParagraphComponent,
} as ComponentMeta<typeof ParagraphComponent>;

const Template: ComponentStory<typeof ParagraphComponent> = (args) => <Table {...args} />;

export const Paragraphs = Template.bind({});
