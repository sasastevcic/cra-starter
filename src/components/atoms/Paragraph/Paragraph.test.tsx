import renderer from 'react-test-renderer';
import { renderWithTheme } from '../../../utils/testsWithTheme';
import Noop from '../Noop';
import Paragraph from './index';

describe('Paragraph', () => {
	describe('Render', () => {
		it('should render', () => {
			const { getByTestId } = renderWithTheme(
				<Paragraph>
					<Noop />
				</Paragraph>,
			);

			const element = getByTestId('Paragraph');

			expect(element).toBeInTheDocument();
		});

		it('should render as span tag', () => {
			const { getByTestId } = renderWithTheme(
				<Paragraph as="span">
					<Noop />
				</Paragraph>,
			);

			const element = getByTestId('Paragraph');

			expect(element.tagName).toBe('SPAN');
		});

		it('should render correctly', () => {
			const tree = renderer
				.create(
					<Paragraph>
						<Noop />
					</Paragraph>,
				)
				.toJSON();

			expect(tree).toMatchSnapshot();
		});
	});
	describe('Content', () => {
		it('should have valid content', () => {
			const { getByTestId } = renderWithTheme(<Paragraph>Paragraph text</Paragraph>);

			const element = getByTestId('Paragraph');

			expect(element).toHaveTextContent('Paragraph text');
		});

		it('should have elements as children', () => {
			const { getByTestId } = renderWithTheme(
				<Paragraph>
					<span>P</span>aragraph
				</Paragraph>,
			);

			const element = getByTestId('Paragraph');
			const child = element.querySelector('span');

			expect(element).toContainElement(child);
		});
	});
});
