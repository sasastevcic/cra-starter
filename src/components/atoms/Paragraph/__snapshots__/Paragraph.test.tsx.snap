// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Paragraph Render should render correctly 1`] = `
<p
  className="sc-bczRLJ jfbakM"
  data-testid="Paragraph"
/>
`;
